### PR TITLE
Sundry Sema fixes

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
I'm not convinced the second change is an improvement. It's sometimes better than no result, but other times we end up with even stranger diagnostics. I'd hold off on it for now.

**Comment by slavapestov** [comment]
Sorry which change? The "invalid pattern" diagnostic change isn't quite done -- I *do* want to clean up that spew of crap in the existing test before I check it in.

**Comment by jrose-apple** [comment]
The name-lookup-filtering change. I think it needs a little more consideration to decide if it really makes sense, given the changes in diagnostics.

**Comment by slavapestov** [comment]
@jrose-apple I think the diagnostics get better -- instead of "cannot find 'foo'" it says "cannot call type member on instance". Also, having a special case *just* for methods (and not properties, etc) seemed a little odd.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test and merge

File: test/Sema/enum_equatable_hashable.swift [hunk 1]
```diff
@@ -96,7 +96,8 @@ private enum Bar<T> {
 
   mutating func value() -> T {
     switch self {
-    case E(let x): // expected-error{{invalid pattern}}
+    // FIXME: Should diagnose here that '.' needs to be inserted, but E has an ErrorType at this point
+    case E(let x):
```
Comment (jrose-apple) [inline test/Sema/enum_equatable_hashable.swift]: We do have to emit *some* error here, or SILGen will crash.

Comment (jrose-apple) [inline test/Sema/enum_equatable_hashable.swift]: Ah, but if it only happens on error cases then we're fine. I see.

File: test/expr/unary/selector/property.swift [hunk 2]
```diff
@@ -37,7 +37,7 @@ class ObjCClass {
     let _ = #selector(setter: myConstant) // expected-error {{argument of '#selector(setter:)' refers to non-settable let 'myConstant'}}
     let _ = #selector(setter: myComputedReadOnlyProperty) // expected-error {{argument of '#selector(setter:)' refers to non-settable var 'myComputedReadOnlyProperty'}}
 
-    let _ = #selector(myClassFunc) // expected-error{{use of unresolved identifier 'myClassFunc'}}
+    let _ = #selector(myClassFunc) // expected-error{{static member 'myClassFunc' cannot be used on instance of type 'ObjCClass'}}
```
Comment (jrose-apple) [inline test/expr/unary/selector/property.swift]: This one is the one that would be nice to fix the error message for.

File: validation-test/IDE/crashers_fixed/069-swift-typechecker-typecheckparameterlist.swift [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// RUN: %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
+// REQUIRES: asserts
```
Comment (jrose-apple) [inline validation-test/IDE/crashers_fixed/069-swift-typechecker-typecheckparameterlist.swift]: This should be removed.

File: validation-test/compiler_crashers_fixed/28376-swift-constraints-constraintsystem-diagnosefailureforexpr.swift [hunk 4]
```diff
@@ -5,7 +5,7 @@
 // See http://swift.org/LICENSE.txt for license information
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: not --crash %target-swift-frontend %s -parse
+// RUN: not %target-swift-frontend %s -parse
 // REQUIRES: asserts
```
Comment (jrose-apple) [inline validation-test/compiler_crashers_fixed/28376-swift-constraints-constraintsystem-diagnosefailureforexpr.swift]: Here too.

**Comment by jrose-apple** [comment]
Slava convinced me in person that the changes for static members are net good, and certainly more consistent.

**Pull request merged**